src/common/sfinae_helpers.h

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+// the loose definitions of types in this file are, well, loose.
+//
+// these helpers aren't here for absolute type certainty at compile-time,
+// but rather to help with templated functions telling types apart.
+
+namespace sfinae
+{
+
+  typedef char true_type;
+
+  struct false_type { true_type a[2]; };
+
+  template <typename T>
+  struct is_not_container
```

Comment:
Why not is_container ? Maybe a bit picky, but less negations might help on the call sites.